### PR TITLE
[opt](invert index) modify of time series compaction policy

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/compaction.cpp [hunk 1]
```diff
@@ -133,10 +133,13 @@ void Compaction::init_profile(const std::string& label) {
 Status Compaction::merge_input_rowsets() {
     std::vector<RowsetReaderSharedPtr> input_rs_readers;
     input_rs_readers.reserve(_input_rowsets.size());
+    int64_t max_compaction_level = 0;
     for (auto& rowset : _input_rowsets) {
         RowsetReaderSharedPtr rs_reader;
         RETURN_IF_ERROR(rowset->create_reader(&rs_reader));
         input_rs_readers.push_back(std::move(rs_reader));
+        max_compaction_level =
```
Comment (csun5285) [inline be/src/olap/compaction.cpp:141]: 每次 compaction 执行时，应该同一层的合并，还是level0 和 level1 的可以混合？

Comment (csun5285) [reply be/src/olap/compaction.cpp:141]: 如果是只能同一层的合并，这里校验下level是否相等？

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 2]
```diff
@@ -70,15 +78,37 @@ uint32_t TimeSeriesCumulativeCompactionPolicy::calc_cumulative_compaction_score(
...
     }
 
-    // Condition 1: the size of input files for compaction meets the requirement of parameter compaction_goal_size
     int64_t compaction_goal_size_mbytes =
             tablet->tablet_meta()->time_series_compaction_goal_size_mbytes();
-    if (total_size >= compaction_goal_size_mbytes * 1024 * 1024) {
+    if (tablet->tablet_meta()->time_series_compaction_level_threshold() == 2) {
+        int32_t rs_meta_count = 0;
+        int64_t continuous_size = 0;
+        for (const auto& rs_meta : checked_rs_metas) {
```
Comment (csun5285) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:86]: 上面 rs_metas 是无序的，checked_rs_metas 里面也是无序的

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 3]
```diff
@@ -47,8 +50,13 @@ uint32_t TimeSeriesCumulativeCompactionPolicy::calc_cumulative_compaction_score(
             continue;
         } else {
             // collect the rowsets of cumulative part
-            total_size += rs_meta->total_disk_size();
             score += rs_meta->get_compaction_score();
+            if (rs_meta->compaction_level() == 0) {
+                level0_total_size += rs_meta->total_disk_size();
+                level0_score += rs_meta->get_compaction_score();
+            } else {
+                checked_rs_metas.push_back(rs_meta);
```
Comment (csun5285) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:58]: 这个函数只是计算下分数，可以在这直接计算 level1 是否满足条件

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 4]
```diff
@@ -193,6 +223,41 @@ int TimeSeriesCumulativeCompactionPolicy::pick_input_rowsets(
...
+    int64_t compaction_goal_size_mbytes =
+            tablet->tablet_meta()->time_series_compaction_goal_size_mbytes();
+
+    std::vector<RowsetSharedPtr> level1_rowsets;
+    if (tablet->tablet_meta()->time_series_compaction_level_threshold() == 2) {
+        int64_t continuous_size = 0;
+        int64_t cumulative_point = tablet->cumulative_layer_point();
+        for (const auto& rowset : candidate_rowsets) {
+            const auto& rs_meta = rowset->rowset_meta();
+            if (rs_meta->end_version() < cumulative_point) {
```
Comment (csun5285) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:235]: 可以删掉，传进来的candidate_rowsets 的版本一定是大于等于cumulative_point的

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 5]
```diff
@@ -193,6 +223,41 @@ int TimeSeriesCumulativeCompactionPolicy::pick_input_rowsets(
...
+                continue;
+            }
+            if (rs_meta->compaction_level() == 0) {
+                break;
+            }
+            int64_t rs_meta_size = rs_meta->total_disk_size();
+            if (rs_meta_size >= compaction_goal_size_mbytes * 2 * 1024 * 1024) {
```
Comment (csun5285) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:242]: 为什么要乘以2

File: be/src/olap/compaction.cpp [hunk 6]
```diff
@@ -133,10 +133,13 @@ void Compaction::init_profile(const std::string& label) {
...
+                std::max(max_compaction_level, rowset->rowset_meta()->compaction_level());
     }
```
Comment (csun5285) [inline be/src/olap/compaction.cpp:143]: 如果所有的输入版本都为空，不需要提升 compation_level

**Review by csun5285 — CHANGES_REQUESTED** [review]

File: be/src/agent/task_worker_pool.cpp [hunk 7]
```diff
@@ -746,6 +746,16 @@ void update_tablet_meta_callback(StorageEngine& engine, const TAgentTaskRequest&
                     tablet_meta_info.time_series_compaction_empty_rowsets_threshold);
             need_to_save = true;
         }
+        if (tablet_meta_info.__isset.time_series_compaction_level_threshold) {
+            if (tablet->tablet_meta()->compaction_policy() != "time_series") {
```
Comment (xiaokang) [inline be/src/agent/task_worker_pool.cpp:750]: change "time_series" literal to a const

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 8]
```diff
@@ -288,7 +349,40 @@ void TimeSeriesCumulativeCompactionPolicy::update_cumulative_point(
...
+    }
+
+    int64_t first_level = 0;
+    for (size_t i = 0; i < input_rowsets.size(); i++) {
+        int64_t cur_level = input_rowsets[i]->rowset_meta()->compaction_level();
+        if (i == 0) {
+            first_level = cur_level;
+        } else {
+            if (first_level != cur_level) {
+                LOG(ERROR) << "Failed to check compaction level, first_level: " << first_level
```
Comment (xiaokang) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:374]: What's the effect of LOG(ERROR)? If it just add a log and continue, it's not enough.

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 9]
```diff
@@ -70,15 +80,32 @@ uint32_t TimeSeriesCumulativeCompactionPolicy::calc_cumulative_compaction_score(
         return score;
     }
 
-    // Condition 1: the size of input files for compaction meets the requirement of parameter compaction_goal_size
     int64_t compaction_goal_size_mbytes =
             tablet->tablet_meta()->time_series_compaction_goal_size_mbytes();
-    if (total_size >= compaction_goal_size_mbytes * 1024 * 1024) {
+    if (tablet->tablet_meta()->time_series_compaction_level_threshold() == 2) {
```
Comment (xiaokang) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:85]: what about 3?

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 10]
```diff
@@ -163,6 +190,13 @@ void TimeSeriesCumulativeCompactionPolicy::calculate_cumulative_point(
             break;
         }
 
+        // upgrade: [0 0 2 1 1 0 0]
+        if (!is_delete && tablet->tablet_meta()->time_series_compaction_level_threshold() == 2 &&
+            rs->compaction_level() == 1) {
```
Comment (xiaokang) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:195]: what about [2, 2, 0, 0, 0]

File: be/src/olap/cumulative_compaction_time_series_policy.cpp [hunk 11]
```diff
@@ -288,7 +349,40 @@ void TimeSeriesCumulativeCompactionPolicy::update_cumulative_point(
...
+        } else {
+            if (first_level != cur_level) {
+                LOG(ERROR) << "Failed to check compaction level, first_level: " << first_level
+                           << ", cur_level: " << cur_level;
+            }
+        }
+    }
+
+    if (output_rowset->rowset_meta()->total_disk_size() >=
+        tablet->tablet_meta()->time_series_compaction_goal_size_mbytes() * 1024 * 1024) {
```
Comment (xiaokang) [inline be/src/olap/cumulative_compaction_time_series_policy.cpp:381]: You can use goal_size * percent (eg. 0.6) to avoid compaction for 0.9 + 0.9

**Review by xiaokang — COMMENTED** [review]
an idea: the priority of level 1 compaction can be lower to give more resource to query and load and level 0 compaction.

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Review by csun5285 — APPROVED** [review]
LGTM

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**